File: kotlin-node/karakum/src/jsMain/resources/nameResolvers/resolveFsOptionsName.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default (node, context) => {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("fs.d.ts") && !sourceFileName.endsWith("fs/promises.d.ts")) return null

    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    let parameter = getParent(node)
    if (ts.isParenthesizedTypeNode(parameter)) {
        parameter = getParent(parameter)
    }
    if (ts.isUnionTypeNode(parameter)) {
        parameter = getParent(parameter)
    }

    if (!parameter) return null
    if (!ts.isParameter(parameter)) return null
    if (!ts.isIdentifier(parameter.name)) return null

    const parameterName = parameter.name.text

    let parentName

    const signatureDeclaration = getParent(parameter)
    if (!signatureDeclaration) return null

    if (ts.isFunctionDeclaration(signatureDeclaration)) {
        if (signatureDeclaration.name === undefined) return null

        parentName = signatureDeclaration.name.text
    } else if (ts.isCallSignatureDeclaration(signatureDeclaration)) {
        const interfaceNode = getParent(signatureDeclaration)
        if (!interfaceNode) return null
        if (!ts.isInterfaceDeclaration(interfaceNode)) return null

        parentName = interfaceNode.name.text
    } else if (ts.isMethodSignature(signatureDeclaration)) {
        if (signatureDeclaration.name === undefined) return null

        const interfaceNode = getParent(signatureDeclaration)
        if (!interfaceNode) return null
        if (!ts.isInterfaceDeclaration(interfaceNode)) return null

        parentName = `${karakum.capitalize(interfaceNode.name.text)}${karakum.capitalize(signatureDeclaration.name.text)}`
    } else {
        return null
    }

    let postfix = ""
    if (
        parameterName === "options"
        && sourceFileName.endsWith("fs/promises.d.ts")
    ) {
        postfix = "Async"
    }

    if (
        parentName === "readdir"
        || parentName === "readdirSync"
        || parentName === "watch"
    ) {
        let infix = "String"

        if (
            ts.isTypeLiteralNode(node)
            && node.members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "encoding"
                && ts.isLiteralTypeNode(member.type)
                && ts.isStringLiteral(member.type.literal)
                && member.type.literal.text === "buffer"
            ))
        ) {
            infix = "Buffer"
        }

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[1]
            && ts.isTypeLiteralNode(node.types[1])
            && node.types[1].members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "encoding"
                && ts.isLiteralTypeNode(member.type)
                && ts.isStringLiteral(member.type.literal)
                && member.type.literal.text === "buffer"
            ))
        ) {
            infix = "Buffer"
        }

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[0]
            && ts.isTypeReferenceNode(node.types[0])
            && ts.isIdentifier(node.types[0].typeName)
            && node.types[0].typeName.text === "ObjectEncodingOptions"
        ) {
            if (
                node.types[1]
                && ts.isTypeLiteralNode(node.types[1])
                && node.types[1].members.some(member => (
                    ts.isPropertySignature(member)
                    && ts.isIdentifier(member.name)
                    && member.name.text === "withFileTypes"
                    && ts.isLiteralTypeNode(member.type)
                    && member.type.literal.kind === ts.SyntaxKind.TrueKeyword
                ))
            ) {
                infix = "WithFileTypes"
            } else {
                infix = ""
            }
        }

        return `${karakum.capitalize(parentName)}${infix}${postfix}${karakum.capitalize(parameterName)}`
    }

    if (
        parentName === "readFile"
        || parentName === "readFileSync"
        || parentName === "FileHandleReadFile"
    ) {
        let infix = "Buffer"

        if (
            ts.isTypeLiteralNode(node)
            && node.members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "encoding"
                && ts.isTypeReferenceNode(member.type)
                && ts.isIdentifier(member.type.typeName)
                && member.type.typeName.text === "BufferEncoding"
            ))
        ) {
            infix = "String"
        }

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[0]
            && ts.isTypeLiteralNode(node.types[0])
            && node.types[0].members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "encoding"
                && ts.isTypeReferenceNode(member.type)
                && ts.isIdentifier(member.type.typeName)
                && member.type.typeName.text === "BufferEncoding"
            ))
        ) {
            infix = "String"
        }

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[0]
            && ts.isTypeReferenceNode(node.types[0])
            && ts.isIdentifier(node.types[0].typeName)
            && node.types[0].typeName.text === "ObjectEncodingOptions"
        ) {
            infix = ""
        }

        return `${karakum.capitalize(parentName)}${infix}${postfix}${karakum.capitalize(parameterName)}`
    }

    if (
        parentName === "stat"
        || parentName === "statSync"
        || parentName === "statfs"
        || parentName === "statfsSync"
        || parentName === "fstat"
        || parentName === "fstatSync"
        || parentName === "lstat"
        || parentName === "watchFile"
        || parentName === "FileHandleStat"
    ) {
        let infix = parentName === "stat" || parentName === "watchFile" ? "Simple" : ""

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[1]
            && ts.isTypeLiteralNode(node.types[1])
            && node.types[1].members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "bigint"
                && ts.isLiteralTypeNode(member.type)
                && member.type.literal.kind === ts.SyntaxKind.TrueKeyword
            ))
        ) {
            infix = "BigInt"
        }

        return `${karakum.capitalize(parentName)}${infix}${postfix}${karakum.capitalize(parameterName)}`
    }

    if (
        parentName === "mkdir"
        || parentName === "mkdirSync"
    ) {
        let infix = ""

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[1]
            && ts.isTypeLiteralNode(node.types[1])
            && node.types[1].members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "recursive"
                && ts.isLiteralTypeNode(member.type)
                && member.type.literal.kind === ts.SyntaxKind.TrueKeyword
            ))
        ) {
            infix = "Recursive"
        }

        return `${karakum.capitalize(parentName)}${infix}${postfix}${karakum.capitalize(parameterName)}`
    }

    if (parentName === "StatSyncFn") {
        let primaryInfix = "Simple"
        let secondaryInfix = "ThrowIfNoEntry"

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[1]
            && ts.isTypeLiteralNode(node.types[1])
            && node.types[1].members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "bigint"
                && ts.isLiteralTypeNode(member.type)
                && member.type.literal.kind === ts.SyntaxKind.TrueKeyword
            ))
        ) {
            primaryInfix = "BigInt"
        }

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[1]
            && ts.isTypeLiteralNode(node.types[1])
            && node.types[1].members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "throwIfNoEntry"
                && ts.isLiteralTypeNode(member.type)
                && member.type.literal.kind === ts.SyntaxKind.FalseKeyword
            ))
        ) {
            secondaryInfix = ""
        }

        if (
            ts.isIntersectionTypeNode(node)
            && node.types[1]
            && ts.isTypeLiteralNode(node.types[1])
            && node.types[1].members.some(member => (
                ts.isPropertySignature(member)
                && ts.isIdentifier(member.name)
                && member.name.text === "throwIfNoEntry"
                && ts.isUnionTypeNode(member.type)
                && member.type.types[0]
                && ts.isLiteralTypeNode(member.type.types[0])
                && member.type.types[0].literal.kind === ts.SyntaxKind.FalseKeyword
            ))
        ) {
            // it is not `throwIfNoEntry` options and they are not simple
            primaryInfix = ""
            secondaryInfix = ""
        }

        return `${karakum.capitalize(parentName)}${primaryInfix}${secondaryInfix}${postfix}${karakum.capitalize(parameterName)}`
    }

    return `${karakum.capitalize(parentName)}${postfix}${karakum.capitalize(parameterName)}`
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/nameResolvers/resolveFsOptionsName.kt
package node.karakum.nameResolvers

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import arrow.core.raise.impure
import arrow.core.raise.nullable
import typescript.Node
import typescript.SyntaxKind
import typescript.asArray
import typescript.isCallSignatureDeclaration
import typescript.isFunctionDeclaration
import typescript.isIdentifier
import typescript.isInterfaceDeclaration
import typescript.isIntersectionTypeNode
import typescript.isLiteralTypeNode
import typescript.isMethodSignature
import typescript.isParameter
import typescript.isParenthesizedTypeNode
import typescript.isPropertySignature
import typescript.isStringLiteral
import typescript.isTypeLiteralNode
import typescript.isTypeReferenceNode
import typescript.isUnionTypeNode

fun resolveFsOptionsName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
    ensure(
        sourceFileName.endsWith("fs.d.ts")
                || sourceFileName.endsWith("fs/promises.d.ts")
    )

    val parameter = ensureNotNull(
        typeScriptService.getParent(node)
            ?.let {
                if (isParenthesizedTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
            ?.let {
                if (isUnionTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
    )
    ensure(isParameter(parameter))

    val parameterNameNode = parameter.name
    ensure(isIdentifier(parameterNameNode))
    val parameterName = parameterNameNode.text

    val signatureDeclaration = ensureNotNull(typeScriptService.getParent(parameter))
    val parentName = ensureNotNull(
        nullable {
            ensure(isFunctionDeclaration(signatureDeclaration))
            signatureDeclaration.name?.text
        } ?: nullable {
            ensure(isCallSignatureDeclaration(signatureDeclaration))

            val interfaceNode = ensureNotNull(typeScriptService.getParent(signatureDeclaration))
            ensure(isInterfaceDeclaration((interfaceNode)))

            interfaceNode.name.text
        } ?: nullable {
            ensure(isMethodSignature(signatureDeclaration))

            val methodNameNode = signatureDeclaration.name
            ensure(isIdentifier(methodNameNode))

            val interfaceNode = ensureNotNull(typeScriptService.getParent(signatureDeclaration))
            ensure(isInterfaceDeclaration((interfaceNode)))

            "${interfaceNode.name.text.replaceFirstChar { it.titlecase() }}${methodNameNode.text.replaceFirstChar { it.titlecase() }}"
        }
    )

    val postfix = if (
        parameterName == "options"
        && sourceFileName.endsWith("fs/promises.d.ts")
    ) {
        "Async"
    } else {
        ""
    }

    nullable {
        ensure(
            parentName == "readdir"
                    || parentName == "readdirSync"
                    || parentName == "watch"
        )

        nullable {
            ensure(isTypeLiteralNode(node))
            ensure(node.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "encoding")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isLiteralTypeNode(propertyType))

                    val propertyTypeLiteral = propertyType.literal
                    ensure(isStringLiteral(propertyTypeLiteral))
                    ensure(propertyTypeLiteral.text == "buffer")
                } != null
            })

            nullable {
                ensure(node.members.asArray().any { member ->
                    nullable {
                        ensure(isPropertySignature(member))

                        val propertyName = member.name
                        ensure(isIdentifier(propertyName))
                        ensure(propertyName.text == "withFileTypes")

                        val propertyType = ensureNotNull(member.type)
                        ensure(isLiteralTypeNode(propertyType))
                        ensure(propertyType.literal.kind == SyntaxKind.TrueKeyword)
                    } != null
                })

                "${parentName.replaceFirstChar { it.titlecase() }}BufferWithFileTypes${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
            } ?: run {
                "${parentName.replaceFirstChar { it.titlecase() }}Buffer${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
            }
        } ?: nullable {
            ensure(isIntersectionTypeNode(node))

            val secondType = ensureNotNull(node.types.asArray().getOrNull( 1))
            ensure(isTypeLiteralNode(secondType))
            ensure(secondType.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "encoding")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isLiteralTypeNode(propertyType))

                    val propertyTypeLiteral = propertyType.literal
                    ensure(isStringLiteral(propertyTypeLiteral))
                    ensure(propertyTypeLiteral.text == "buffer")
                } != null
            })

            "${parentName.replaceFirstChar { it.titlecase() }}Buffer${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: nullable {
            ensure(isIntersectionTypeNode(node))

            val firstType = ensureNotNull(node.types.asArray().first())
            ensure(isTypeReferenceNode(firstType))

            val typeName = firstType.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "ObjectEncodingOptions")

            nullable {
                val secondType = ensureNotNull(node.types.asArray().getOrNull( 1))
                ensure(isTypeLiteralNode(secondType))
                ensure(secondType.members.asArray().any { member ->
                    nullable {
                        ensure(isPropertySignature(member))

                        val propertyName = member.name
                        ensure(isIdentifier(propertyName))
                        ensure(propertyName.text == "withFileTypes")

                        val propertyType = ensureNotNull(member.type)
                        ensure(isLiteralTypeNode(propertyType))
                        ensure(propertyType.literal.kind == SyntaxKind.TrueKeyword)
                    } != null
                })

                "${parentName.replaceFirstChar { it.titlecase() }}WithFileTypes${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
            } ?: run {
                "${parentName.replaceFirstChar { it.titlecase() }}${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
            }
        } ?: run {
            "${parentName.replaceFirstChar { it.titlecase() }}String${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        }
    } ?: nullable {
        ensure(
            parentName == "readFile"
                    || parentName == "readFileSync"
                    || parentName == "FileHandleReadFile"
        )

        nullable {
            ensure(isTypeLiteralNode(node))
            ensure(node.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "encoding")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isTypeReferenceNode(propertyType))

                    val typeName = propertyType.typeName
                    ensure(isIdentifier(typeName))
                    ensure(typeName.text == "BufferEncoding")
                } != null
            })

            "${parentName.replaceFirstChar { it.titlecase() }}String${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: nullable {
            ensure(isIntersectionTypeNode(node))

            val firstType = ensureNotNull(node.types.asArray().first())
            ensure(isTypeLiteralNode(firstType))
            ensure(firstType.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "encoding")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isTypeReferenceNode(propertyType))

                    val typeName = propertyType.typeName
                    ensure(isIdentifier(typeName))
                    ensure(typeName.text == "BufferEncoding")
                } != null
            })

            "${parentName.replaceFirstChar { it.titlecase() }}String${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: nullable {
            ensure(isIntersectionTypeNode(node))

            val firstType = ensureNotNull(node.types.asArray().first())
            ensure(isTypeReferenceNode(firstType))

            val typeName = firstType.typeName
            ensure(isIdentifier(typeName))
            ensure(typeName.text == "ObjectEncodingOptions")

            "${parentName.replaceFirstChar { it.titlecase() }}${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: run {
            "${parentName.replaceFirstChar { it.titlecase() }}Buffer${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        }
    } ?: nullable {
        ensure(
            parentName == "stat"
                    || parentName == "statSync"
                    || parentName == "statfs"
                    || parentName == "statfsSync"
                    || parentName == "fstat"
                    || parentName == "fstatSync"
                    || parentName == "lstat"
                    || parentName == "watchFile"
                    || parentName == "FileHandleStat"
        )

        nullable {
            ensure(isIntersectionTypeNode(node))

            val secondType = ensureNotNull(node.types.asArray().getOrNull( 1))
            ensure(isTypeLiteralNode(secondType))
            ensure(secondType.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "bigint")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isLiteralTypeNode(propertyType))
                    ensure(propertyType.literal.kind == SyntaxKind.TrueKeyword)
                } != null
            })

            "${parentName.replaceFirstChar { it.titlecase() }}BigInt${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: nullable {
            ensure(parentName == "stat" || parentName == "watchFile")

            "${parentName.replaceFirstChar { it.titlecase() }}Simple${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: run {
            "${parentName.replaceFirstChar { it.titlecase() }}${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        }
    } ?: nullable {
        ensure(
            parentName == "mkdir"
                    || parentName == "mkdirSync"
        )

        nullable {
            ensure(isIntersectionTypeNode(node))

            val secondType = ensureNotNull(node.types.asArray().getOrNull( 1))
            ensure(isTypeLiteralNode(secondType))
            ensure(secondType.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "recursive")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isLiteralTypeNode(propertyType))
                    ensure(propertyType.literal.kind == SyntaxKind.TrueKeyword)
                } != null
            })

            "${parentName.replaceFirstChar { it.titlecase() }}Recursive${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        } ?: run {
            "${parentName.replaceFirstChar { it.titlecase() }}${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
        }
    } ?: nullable {
        ensure(parentName == "StatSyncFn")

        var primaryInfix = "Simple"
        var secondaryInfix = "ThrowIfNoEntry"

        impure {
            ensure(isIntersectionTypeNode(node))

            val secondType = ensureNotNull(node.types.asArray().getOrNull( 1))
            ensure(isTypeLiteralNode(secondType))
            ensure(secondType.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "bigint")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isLiteralTypeNode(propertyType))
                    ensure(propertyType.literal.kind == SyntaxKind.TrueKeyword)
                } != null
            })

            primaryInfix = "BigInt"
        }

        impure {
            ensure(isIntersectionTypeNode(node))

            val secondType = ensureNotNull(node.types.asArray().getOrNull( 1))
            ensure(isTypeLiteralNode(secondType))
            ensure(secondType.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "throwIfNoEntry")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isLiteralTypeNode(propertyType))
                    ensure(propertyType.literal.kind == SyntaxKind.FalseKeyword)
                } != null
            })

            secondaryInfix = ""
        }

        impure {
            ensure(isIntersectionTypeNode(node))

            val secondType = ensureNotNull(node.types.asArray().getOrNull( 1))
            ensure(isTypeLiteralNode(secondType))
            ensure(secondType.members.asArray().any { member ->
                nullable {
                    ensure(isPropertySignature(member))

                    val propertyName = member.name
                    ensure(isIdentifier(propertyName))
                    ensure(propertyName.text == "throwIfNoEntry")

                    val propertyType = ensureNotNull(member.type)
                    ensure(isUnionTypeNode(propertyType))

                    val firstPropertyType = ensureNotNull(propertyType.types.asArray().first())
                    ensure(isLiteralTypeNode(firstPropertyType))
                    ensure(firstPropertyType.literal.kind == SyntaxKind.FalseKeyword)
                } != null
            })

            // it is not `throwIfNoEntry` options and they are not simple
            primaryInfix = ""
            secondaryInfix = ""
        }

        "${parentName.replaceFirstChar { it.titlecase() }}${primaryInfix}${secondaryInfix}${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
    } ?: run {
        "${parentName.replaceFirstChar { it.titlecase() }}${postfix}${parameterName.replaceFirstChar { it.titlecase() }}"
    }
}
